### ISSUE TITLE
The loss doesn't seem to converge after 1000 iterations

### ISSUE THREAD

**Issue by Yiltan** [karpathy/llm.c#138]
I'm working on the C version of the code in preparation for (#40)

So llm.c with **no** code modifications I observe the following:
- `test_gpt2` works successfully and the loss matches
- `train_gpt2` after 1000 iterations the loss sill hovers around `4.0`
    - I notice similar behaviour for `40` iterations on the README
    
I'm curious if this a platform specific bug (I'm using Intel Xeons rather than ARM) or something with the code itself. Also GPT2 may take many iterations and I need to wait longer. Any guidance would be useful.

**Comment by karpathy** [comment]
I think it's because the first few iterations are super easy gains and after that they get much less easy. Also the dataset is super tiny, so it may be that one cannot do much better on the val loss. The train loss should continue to improve though, if you smooth it enough, or evaluate it over a large set of batches and average. Also note that default training on non .cu version is tiny - batch 4, seq len 64.

**Comment by dagelf** [comment]
I see other implementations train for 10000's of steps and on other base models to get lower losses on `tinyshakespeare` and `tinystories`. Bunch of interesting papers. 

You compare the outputs of different `CFLAGS`, might make minor differences. (also add `-march=native`)

**Comment by Yiltan** [comment]
Both of your advice was quite useful. I ran it a little more an the losses eventually converged. 

However, validation loss >> training loss, which suggests over fitting.

- Do you have any suggestions for larger data sets?
- Are there any training parameters I can adjust?

![llm c_train](https://github.com/karpathy/llm.c/assets/9093579/06aa2337-f809-44bb-8333-9ef36fee1590)

As a side note, the distributed MPI implementation is pretty much done.
